### ISSUE TITLE
Bug on created_at in customer_entity table on update

### ISSUE THREAD

**Issue by sdecalom** [OpenMage/magento-lts#4069]
### Preconditions (*)

1.OpenMage 19.4.23
2.Mysql 8.0

### Steps to reproduce (*)

Update a customer created on 2024-07-01 at 15:00.

### Expected result (*)

The date of created_at on customer_entity should be 2024.07.01 15:00
![image](https://github.com/OpenMage/magento-lts/assets/3275377/e282243a-323e-4997-9c1b-dbd6a5304146)

### Actual result (*)

The created_at date on customer_entity is 2024.07.01 17:00
![image](https://github.com/OpenMage/magento-lts/assets/3275377/bf8c448f-4643-4cf3-bfa2-4d1b4aa6773a)

**Comment by ADDISON74** [comment]
You have a very old version, unmaintained. 

I confirm with the cloned version of the repository from today I cannot reproduce it. I created a customer account in the Backend and the created_at column shows correct information. I updated an existing customer then the column changes with the correct timestamp.

Please try reproducing this issue in a test environment just with the latest version without any extension, modules, or custom work.

**Comment by sdecalom** [comment]
OK thank you for your support, I will try with the latest version.

**Comment by sdecalom** [comment]
OK thank you for your support, I will try with the latest version.

**Comment by ADDISON74** [comment]
Did you manage to solve this issue? 

In my tests I used MariaDB and Percona, but I will try to MySQL 8 in order to reproduce this issue.

**Comment by sdecalom** [comment]
Hello,

I manage this issue with this pull request:
https://github.com/OpenMage/magento-lts/pull/4070


________________________________
De : ADDISON ***@***.***>
Envoyé : mercredi 3 juillet 2024 17:25
À : OpenMage/magento-lts ***@***.***>
Cc : Sébastien DECALOM ***@***.***>; State change ***@***.***>
Objet : Re: [OpenMage/magento-lts] Bug on created_at in customer_entity table on update (Issue #4069)


Did you manage to solve this issue?

In my tests I used MariaDB and Percona, but I will try to MySQL 8 in order to reproduce this issue.

**Comment by ma4nn** [comment]
Why is the `created_at` column on the `customer_entity` table set to `on update current_timestamp()` at all? 
I noticed this coincidentally after doing a manual `UPDATE` on the table and now all customer created dates are overwritten. This must be a bug, right?

**Comment by sreichel** [comment]
@ma4nn its not only customer_entity ... its for all entity tables. Seems not correct to me. `updated_at` should be updated.

**Comment by ma4nn** [comment]
@sreichel ah yes you're right, even worse 👍 any idea why this has been done? I also don't get why nobody complained about that yet? It's a massive bug and the data cannot be recovered at all (or at least it's an immense effort depending on the backups) if you didn't notice immediately.

Sure it's only relevant if you do direct sql updates, but this should not be so uncommon I think..

**Comment by sreichel** [comment]
> I also don't get why nobody complained about that yet?

Using models work and the most used, product and category are correct?

Should be fixed ....

Suggestion for all?

```
    created_at        timestamp       default current_timestamp()   not null comment 'Created At',
    updated_at        timestamp       default '0000-00-00 00:00:00' not null on update current_timestamp() comment 'Updated At',
```

**Comment by justinbeaty** [comment]
I guess everyone is using the sample data. 😅 Without it:

```
  `created_at` timestamp NOT NULL DEFAULT '0000-00-00 00:00:00' COMMENT 'Created At',
  `updated_at` timestamp NOT NULL DEFAULT '0000-00-00 00:00:00' COMMENT 'Updated At',
```

**Comment by sreichel** [comment]
Removing current timestamp for created_at could break import-scripts expected behavoir. (?)

**Comment by justinbeaty** [comment]
@sreichel What I'm saying is that without installing the sample data, that is how the columns are defined.

**Comment by justinbeaty** [comment]
However, I'm sure it would be better to be:

```
  `created_at` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP COMMENT 'Created At',
  `updated_at` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP COMMENT 'Updated At',
```

Then entirely remove the two files at `app/code/core/Mage/Eav/Model/Entity/Attribute/Backend/Time/`, and just convert to store timezone in an `_afterLoad` somewhere, maybe `Mage_Eav_Model_Entity_Attribute`?

**Comment by sreichel** [comment]
> @sreichel What I'm saying is that without installing the sample data, that is how the columns are defined.

Ahhhh.  Have seen it now. :rofl: 

So "only" sample data needs to be updated for https://github.com/OpenMage/magento-lts/issues/4069#issuecomment-2453376772? After years it needs also some other updates ...

> However, I'm sure it would be better to be

:+1:

**Comment by justinbeaty** [comment]
> So "only" sample data needs to be updated for https://github.com/OpenMage/magento-lts/issues/4069#issuecomment-2453376772? After years it needs also some other updates ...

Normally sample data shouldn't need to be updated for schema changes, because upgrade scripts would take care of that. So how did it happen? Did the author of the sample data just change the schema manually!?

This original issue and #4070 still seem valid, but my idea in the last comment could solve it.

**Comment by justinbeaty** [comment]
@ma4nn Yeah it's definitely odd because as far back as [Magento 1.6.0.0-alpha1](https://github.com/OpenMage/magento-lts/blame/a75e8625f439392a1d0cb113aa93223aa1d1a58a/app/code/core/Mage/Customer/sql/customer_setup/install-1.6.0.0.php#L62) it never had the ON UPDATE statement.

Also, Varien has some constants to help up with this:
```php
    /**
     * Default values for timestampses - fill with current timestamp on inserting record, on changing and both cases
     */
    public const TIMESTAMP_INIT_UPDATE = 'TIMESTAMP_INIT_UPDATE';
    public const TIMESTAMP_INIT        = 'TIMESTAMP_INIT';
    public const TIMESTAMP_UPDATE      = 'TIMESTAMP_UPDATE';
```

**Comment by sreichel** [comment]
I will update sample-data ASAP. (nearly done)

Changing install scripts is another task.

@justinbeaty thanks!

**Comment by sreichel** [comment]
> Changing install scripts is another task.

@justinbeaty can you please add your PR, to deleted outdated scripts.
